Removed ~50% of metaprogramming from LibraryEvalContext & Resource

zenspider opened the pull request:
This should probably be reviewed by commit.

A majority of that file count is BS as I nuked AWS tests just to get their noise out of the way. I can rebase that out and get them passing if we actually care. I also put up a PR to remove AWS + azure for funsies.

zenspider committed:
Completely refactored profile_resource_exceptions_test.rb

Half... half.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Add test for LEC to expose second-level libraries require.

Not the best test in the world, but it is better than nothing.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Improve test expressiveness by using more direct assertions.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Minor test improvements

We need to track down `x.must_equal true`.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Don't meta-program if you don't need to.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Clean up Platform#params.

State your normal case first, then deal with edge cases, not the other
way around.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Overhaul Platform#supported?

+ State normal path first.
+ Use Enumerable to state your logic up front, don't build it into
  loops with breaks/returns.
+ Remove nonsensical duplication of key formats. (should push up or remove)
+ Had to add an extra test case that wasn't covered here but was in
  metadata_test.rb (why?!?)

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
These tests don't pass on a network cafe and just hang forever.

Changing to localhost resolves immediately but assumes you're NOT
running a git server on http locally.

This seems more valid to me than assuming you know how DNS is going to
resolve everywhere.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Clean up the end of our custom require.

There are only 2 options. Shortcutting one of them with a `require` just
makes it messy. Use `if`. It's really basic. Basic is good. Maybe we
should try to push towards basic and good. It would be work but make
us happier in the long run.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Mostly minor cleanup, clarification, and comments.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Added extra doco on our require method.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
First pass reworking LibraryEvalContext.

This takes 3 anonymous Class.new w/ no superclasses, no documentation,
and no obvious clue as to what anything is and drops it down to 2
Class.new w/ superclasses, documents what it can, and makes it much
more clear what is going on. (yet, still not clear)

It also actually uses LibraryEvalContext. Having names is important.
Probably more important than any of the rest of this diff.

One extra thing this adds is Module.find_class_instance_variable and
hooks it up to Resource.__resource_registry. This allows it to find a
class-level registry regardless of whether it was hidden in an
anonymous superclass or not.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Simple refactoring + assign __resource_registry to the current if needed.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
This is all dead. Removed.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Rename Inspec::Resource.supports to support_registry.

To fix an already existing name collision between Inspec::Resource and
Inspec::Plugins::Resource.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Merged Inspec::Plugins::Resource to Inspec::Resource.

This also removes Inspec::ResourceBehaviors and Inspec::ResourceDSL.

All class methods look like class methods.
All instance methods look like instance methods.

There is still too much clever going on with __register creating
another anonymous subclass, but I think I can work that out/down too.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Reworked Resource.__register's metaprogramming to be as small as possible.

Mostly involved pushing up all the regular initialize code to a
regular method, calling that, THEN doing super up to the resource
initialize.

This brings __register's Class.new to 5 lines.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Moved __register down to bookkeeping section.

Signed-off-by: Ryan Davis <zenspider@chef.io>

clintoncwolfe reviewed (dismissed):
Can we get a clean diff on this, without the unrelated AWS/Azure test deletions?

zenspider committed:
Tweak for aws resources to determine if they're still under test.

Everyone now has an `inspec` method, but the AWS ones still return
nil as their backend hasn't been set up.

This seems wrong...

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Removed metaprogramming in Inspec::Backend.

Signed-off-by: Ryan Davis <zenspider@chef.io>

miah reviewed (approved):
I paired with @zenspider for like 98% of this. :+1:

clintoncwolfe commented on test/unit/profiles/profile_resource_exceptions_test.rb:
```diff
@@ -6,107 +6,68 @@
     profile = MockLoader.load_profile("profile-with-resource-exceptions")
     profile.load_libraries
-    profile.collect_tests
+    profile.collect_tests # TODO: bad name?
```
Agreed, I've always thought that was a bad name. In our online docs, we have profile -> control -> describe block -> test (`its` block). I *think* here `collect_tests` is actually referring to `describe_blocks`/ExampleGroups.

clintoncwolfe commented on lib/inspec/resources/platform.rb:
```diff
@@ -71,31 +71,26 @@ def params
+            check_release(v)
+          end
+        }
+      } || supports.empty?
```
The string/symbol normalization is both regrettable and an improvement :-)

But for the logic, I think this loses a lot of clarity - keep in mind other people who are not you may need to read and understand this. Imagine a junior engineer reading the original logic versus this. What is that dangling `||supports.empty?` doing? Easily missed. 

The original is much more straightforward and less clever.  I think we want code that is easy to maintain by a broad range of engineers.

zenspider replied:
> The original is much more straightforward and less clever

I wholeheartedly disagree. Using enumerable methods spells out intent much more clearly than managing logic via local variables w/ `next`, `break` and the like. As a result using higher level intention describing methods is more junior developer friendly than the original version. Objectively, the new code is 83% the complexity of the original (17.5/20.9).

As far as the hanging `empty?` clause, it is lamentable, but always I prefer to put the positive/main cases before the negative/edge cases whenever possible. I could try to format it slightly differently but I prefer this structure.

clintoncwolfe commented on lib/inspec/dsl_shared.rb:
```diff
@@ -8,6 +8,26 @@ module RequireOverride
+      #   -> load_library_file(@library_eval_context)
+      #
+      # probably most of this comment is useless, but it was hard to
+      # discover so I'm adding it for others.
```
Thank you for this.

clintoncwolfe commented on test/docker_test.rb:
```diff
@@ -1,55 +0,0 @@
-require_relative "docker_run"
```
We do still create a docker image during release - and some people are consuming it, apparently (we had a recent issue that mentioned it indirectly).

zenspider replied:
Sure. But we don't run any of these `test/docker*` files anywhere that I can find.

clintoncwolfe reviewed (commented):
I think this is wonderful - I especially appreciate all of the code comments reflecting what was learned along the way.

I don't have any requested changes - the only thing holding me back from an approval is that since this makes changes to the Resource base class as well as some of the logic behind the platform support, it would be advisable to do another run against the compliance-profile repo (this branch against their master).

clintoncwolfe reviewed (commented):
Hrm, see prior comment review.

james-stocks commented on lib/inspec/control_eval_context.rb:
```diff
@@ -12,236 +12,168 @@ module Inspec
-        define_method :add_subcontext do |context|
-          profile_context_owner.add_subcontext(context)
-        end
+      # TODO: seriously consider getting rid of the NPM model
```
NPM?

zenspider committed:
Refactored and moved Resource.create_dsl to ProfileContext.

ProfileContext#to_resources_dsl is the only usage of it.
The responsibility for this code should be on ProfileContext.

Refactored a majority of #create_dsl to be a 2-line long Module.new
instead of 36 lines. Much less magic. Still has an anonymous module
but it is now much easier to address and debug.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Improved the unit/profiles/* tests.

Use better assertions for better failure messages.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed error on ruby 2.4 because define_method was private.

Signed-off-by: Ryan Davis <zenspider@chef.io>

stevendanna commented on lib/inspec/resources/platform.rb:
```diff
@@ -51,46 +56,41 @@ def in_family?(family)
-        end
-        return true if status == true
-      end
+      raise ArgumentError, "`supports` is nil." unless supports
```
I'm mildly worried raising here rather than returning true will break someone.

stevendanna reviewed (commented):
Your perseverance on this is admirable.  I reviewed this simply because I try to keep a vague sense of how this code base works. A few general comments:

- There are a bunch of small cleanups here that are completely uncontroversial. I'd say if we can't merge this PR soon, it is worth spending a few minutes, pulling those out and getting them merged.

- I'll be honest, this code base is in a state where it is hard for me to judge whether this change to the metaprogramming is making things better or worse.  My guess is that I had just grown familiar to the old structure and that since it is still VERY complex the new structure means I lose the benefit of familiarity.  I trust y'all to do whatever seems like the right thing.

- I think the best thing we could do for ourselves is to write a big piece of documentation somewhere that covers: (1) The various evaluation contexts, what they are used for, and the key points of how they are constructed and (2) the invariants that we _think_ the current system provides with respect to resource registry "isolation"

clintoncwolfe reviewed (approved):
Thanks for all your work on this!
